### PR TITLE
 #184361216 Feat:  Admin should be able to set roles and permissions

### PR DESCRIPTION
**Pull request by ISHIMWE-Aime**
# Description

I have add about four (4) features but all of them concerning about setting roles and permissions; 
those features are: Admin set role to a particular user, Admin set Role and permission in system, Admin add permission(s) to a particular user, Admin remove a permission(s) to a particular user.

NB: what I have not yet done is unity testing and integration test. I am planning to do them, after your suggestions and changes.

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

# Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or Enhancement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have created a loom video for the new feature or enhancement
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/controllers/RoleAndPermission.controller.ts [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    }
+  }
+};
+
+// Admin set permissions w.r.t role
+export const setRoleAndPermission = async (req: Request, res: Response) => {
+  try {
+    const roleAndPermissonInfo = req.body as RoleAndPermissionAttribute;
+
+    if (await getRoleAndPermisson(req.body.role)) {
```
Comment (shyakadavis) [inline src/controllers/RoleAndPermission.controller.ts:59]: May you first sanitize this. 🗑️ 

As it is, you don't know the shape of this input. e.g whether it's a boolean, number, string...

**Review by musllim — COMMENTED** [review]
LGTM

File: src/controllers/RoleAndPermission.controller.ts (hunk 1 above)
Comment (ISHIMWE-Aime) [reply src/controllers/RoleAndPermission.controller.ts]: I have added the validation to all input data I am capturing through those 4 routes I have used. 
and thank you to your valuable feedback, @musllim and @shyakadavis and I hope you are going to review again.

**Review by musllim — APPROVED** [review]
looks good

File: src/routes/roleAndPermission.routes.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+const RM_PERMISSION = process.env.RM_PERMISSION as string;
+const ADD_PERMISSION = process.env.ADD_PERMISSION as string;
+const SET_ROLE = process.env.SET_ROLE as string;
+
+const setRoleAndPermissionRouter = Router();
+
+setRoleAndPermissionRouter.patch(
+  "/update/:id",
+  ValidateJoi(RoleAndPermissionsSchema.roleTobeSet),
+  [protectRoute, permisionCheck(SET_ROLE)],
```
Comment (shyakadavis) [inline src/routes/roleAndPermission.routes.ts:23]: Validation should come after authentication, right?

I think the validation middleware should come as the last check in the process.

Comment (shyakadavis) [inline src/routes/roleAndPermission.routes.ts:23]:
```suggestion
  [protectRoute, permisionCheck(SET_ROLE), ValidateJoi(RoleAndPermissionsSchema.roleTobeSet)],
```

Comment (shyakadavis) [reply src/routes/roleAndPermission.routes.ts:23]: I think the same applies to other routes.

**Pull request closed without merging**